### PR TITLE
Add HttpSecurity.addFilterAt

### PR DESCRIPTION
**Pull request by rwinch**
Fixes gh-3784

### CODE REVIEW

File: config/src/main/java/org/springframework/security/config/annotation/web/builders/FilterComparator.java [hunk 1]
```diff
@@ -150,6 +150,23 @@ public void registerAfter(Class<? extends Filter> filter,
 	}
 
 	/**
+	 * Registers a {@link Filter} to exist at a particular {@link Filter} position
+	 * @param filter the {@link Filter} to register
+	 * @param atFilter the {@link Filter} that is already registered and that
+	 * {@code filter} should be placed at.
+	 */
+	public void registerAt(Class<? extends Filter> filter,
```
Comment (jgrandja) [inline config/src/main/java/org/springframework/security/config/annotation/web/builders/FilterComparator.java]: The new registerAt method has the same implementation as the existing registerAfter method. Possible case for reuse.

File: config/src/main/java/org/springframework/security/config/annotation/web/builders/HttpSecurity.java [hunk 2]
```diff
@@ -1026,6 +1026,26 @@ public HttpSecurity addFilter(Filter filter) {
...
+	 * </pre>
+	 *
+	 * @param filter the Filter to register
+	 * @param atFilter the location of another {@link Filter} that is already registered
+	 * (i.e. known) with Spring Security.
+	 * @return the {@link HttpSecurity} for further customizations
+	 */
+	public HttpSecurity addFilterAt(Filter filter, Class<? extends Filter> atFilter) {
+		this.comparitor.registerAt(filter.getClass(), atFilter);
+		this.filters.add(filter);
```
Comment (jgrandja) [inline config/src/main/java/org/springframework/security/config/annotation/web/builders/HttpSecurity.java]: Both addFilterBefore() and addFilterAfter() call the addFilter() method to add the filter which also does a pre-check before adding. However, the new addFilterAt() method directly puts the filter in the filters member variable avoiding the extra check in addFilter(). Although the first call to comparitor.registerAt() (if it succeeds) would ensure that the addFilter check will succeed too. So the current implementation would work the same way. But it's not consistent with the other 2 methods of adding. Is there a reasoning behind this that I'm not seeing?

File: config/src/main/java/org/springframework/security/config/annotation/web/builders/HttpSecurity.java [hunk 3]
```diff
@@ -1026,6 +1026,26 @@ public HttpSecurity addFilter(Filter filter) {
...
+	 * addFilterAt(new CustomFilter(), UsernamePasswordAuthenticationFilter.class)
+	 * </pre>
+	 *
+	 * @param filter the Filter to register
+	 * @param atFilter the location of another {@link Filter} that is already registered
+	 * (i.e. known) with Spring Security.
+	 * @return the {@link HttpSecurity} for further customizations
+	 */
+	public HttpSecurity addFilterAt(Filter filter, Class<? extends Filter> atFilter) {
+		this.comparitor.registerAt(filter.getClass(), atFilter);
```
Comment (jgrandja) [inline config/src/main/java/org/springframework/security/config/annotation/web/builders/HttpSecurity.java]: Typo on comparitor....should be comparator

File: config/src/main/java/org/springframework/security/config/annotation/web/builders/HttpSecurity.java (hunk 2 above)
Comment (rwinch) [inline config/src/main/java/org/springframework/security/config/annotation/web/builders/HttpSecurity.java]: The check probaly isn't necessary, but I think it reads better and is consistent. I will update this

File: config/src/main/java/org/springframework/security/config/annotation/web/builders/HttpSecurity.java (hunk 3 above)
Comment (rwinch) [inline config/src/main/java/org/springframework/security/config/annotation/web/builders/HttpSecurity.java]: This is a separate issue as comparitor already existed

File: config/src/main/java/org/springframework/security/config/annotation/web/builders/FilterComparator.java (hunk 1 above)
Comment (rwinch) [inline config/src/main/java/org/springframework/security/config/annotation/web/builders/FilterComparator.java]: This was a copy / paste error. The index should be `position` not `position + 1`. This is now fixed

Comment (jgrandja) [inline config/src/main/java/org/springframework/security/config/annotation/web/builders/FilterComparator.java]: Give this logic, if existing FilterA is at position 101
then the call registerAt(FilterB, FilterA)
would result in both filters at position 101
can you see any possible side effects of them being equal?

Comment (rwinch) [inline config/src/main/java/org/springframework/security/config/annotation/web/builders/FilterComparator.java]: Yes, but this is up to the users discretion. This is no different than a user stating to put a key / value pair in a Map if the key already exists

Comment (jgrandja) [inline config/src/main/java/org/springframework/security/config/annotation/web/builders/FilterComparator.java]: Ok sounds good just wanted to point that out.

**Pull request merged**